WalletWasabi/Stores/SqliteStorage.cs

```diff
@@ -0,0 +1,241 @@
+	{
+		return Connection.BeginTransaction();
+	}
+
+	/// <summary>
+	/// Returns all filters with height ≥ than the given one.
+	/// </summary>
+	public IEnumerable<FilterModel> Fetch(int fromHeight)
+	{
+		using SqliteCommand command = Connection.CreateCommand();
+		command.CommandText = "SELECT * FROM filter WHERE block_height >= $block_height ORDER BY block_height";
+		command.Parameters.AddWithValue("$block_height", fromHeight);
+
+		using SqliteDataReader reader = command.ExecuteReader();
+
+		while (reader.Read())
+		{
+			FilterModel filter = ReadRow(reader);
+			yield return filter;
+		}
+	}
+
+	/// <summary>
+	/// Returns last <paramref name="n"/> filters from the database table.
+	/// </summary>
+	public IEnumerable<FilterModel> FetchLast(int n)
+	{
+		using SqliteCommand command = Connection.CreateCommand();
+		command.CommandText = "SELECT * FROM filter WHERE block_height >= (SELECT MAX(block_height) - $n FROM filter) ORDER BY block_height";
+		command.Parameters.AddWithValue("$n", n);
+
+		using SqliteDataReader reader = command.ExecuteReader();
+
+		while (reader.Read())
+		{
+			FilterModel filter = ReadRow(reader);
+			yield return filter;
+		}
+	}
+
+	/// <summary>
+	/// Removes the filter with the highest height from the database table.
+	/// </summary>
+	public bool TryRemoveLast([NotNullWhen(true)] out FilterModel? filter)
+	{
+		using SqliteCommand command = Connection.CreateCommand();
+		command.CommandText = "DELETE FROM filter WHERE block_height = (SELECT MAX(block_height) FROM filter) returning *";
+
+		using SqliteDataReader reader = command.ExecuteReader();
+
+		if (!reader.Read())
+		{
+			filter = null;
+			return false;
+		}
+
+		// Stored filters are considered to be valid.
+		filter = ReadRow(reader);
+		return true;
+	}
+
+	/// <summary>
+	/// Removes the filter with the highest height from the table if the last filter has higher height than <paramref name="height"/>.
+	/// </summary>
+	/// <param name="height">Minimum block height of the last block to remove (exclusive).</param>
+	public bool TryRemoveLastIfNewerThan(uint height, [NotNullWhen(true)] out FilterModel? filter)
+	{
+		using SqliteCommand command = Connection.CreateCommand();
+		command.CommandText = "DELETE FROM filter WHERE block_height > $block_height AND block_height = (SELECT MAX(block_height) FROM filter) returning *";
+		command.Parameters.AddWithValue("$block_height", height);
+
+		using SqliteDataReader reader = command.ExecuteReader();
+
+		if (!reader.Read())
+		{
+			filter = null;
+			return false;
+		}
+
+		// Stored filters are considered to be valid.
+		filter = ReadRow(reader);
+		return true;
+	}
+
+	private FilterModel ReadRow(SqliteDataReader reader)
+	{
+		uint blockHeight = (uint)reader.GetInt64(ordinal: 0);
+		uint256 blockHash = new((byte[])reader["block_hash"]);
+		byte[] filterData = (byte[])reader["filter_data"];
+		uint256 prevBlockHash = new((byte[])reader["previous_block_hash"]);
```

Comment:
I would suggest not to use names when there are ordinals. Otherwise, you resolve an ordinal each row. In worst case, when position of columns isn't known for some reason or there should be no hard dependency on it, ordinals should be resolved before reading any row.

Another rule of thumb is to avoid indexers because they always box and under the hood they always use `GetXxx` methods. Ugliness of API in ADO.NET happened due to the initial design which later changed to a better one and because of backward compatibility. For the same reason there's an interface and a base class for each ADO.NET object while only one thing should be enough.

Reply:
Right, ordinals are more preferred.

A small note, you can use `GetFieldValue<byte[]>` since the reader type is sealed and optimized in such a way, that there should be no generic virtual call. `GetValue` is fine too in case of SQLite because for `byte[]` it is used under the hood by `GetFieldValue`.

Reply:
In case of SQLite it's fine to keep it as is (it is an in process database), but the general practice is disposal of a connection right after finishing all commands in the method. Why so? There are two reasons:

1. A physical connection can be shared thankfully for pooling. That's useful when you have more than one place accessing the same database and potentially from different threads surely if both of them are not accessing it simultaneously.

2. For out of process databases there's a limited number of connections, so a better option here is to limit physical connections made by clients. A good example is PostgreSQL which starts an instance for each incoming connection and stops it when the connection closes. It's a very expensive operation here. Moreover, the listening server has only 100 parallel connections by default.

Sure, we won't use any out of process database, but you said that you are going to use SQLite in another place, so reason number one still applies even if it's partly valid due to the nature of SQLite.